hypothesis/lms
# Support launching assignments using VitalSource books

This commit adds support for loading the VitalSource book viewer in an LTI launch if the assignment URL contains `vitalsource_enabled=true` in the query params, similar to how Canvas files assignments are configured.

The VitalSource book viewer is loaded via a secondary LTI 1.1 launch. When an assignment is configured to use VitalSource, the backend generates form parameters for an LTI launch and signs them using a key and secret specified by two new env vars: `VITALSOURCE_LAUNCH_KEY` and `VITALSOURCE_LAUNCH_SECRET`. These form parameters are rendered into the configuration for the LMS frontend.

When this configuration exists, the LMS frontend creates an empty iframe, renders a form with these parameters into the iframe and auto-submits it. As a result of the submission, the VitalSource book viewer app will load in the iframe at the specified book and chapter with Hypothesis embedded.

There is currently an issue that the VS app always tries to load the production Hypothesis client, which won't work properly with the development LMS app.

Changes in this PR in detail:

 - Whitelist the `VITALSOURCE_LAUNCH_KEY` and `VITALSOURCE_LAUNCH_SECRET` env vars in tox.ini.
 - Add a `VitalSourceService` service in `lms/services/vitalsource.py` to generate the form parameters for the VS LTI launch
 - Add a new LTI launch type, guarded by a `vitalsource_book` predicate, which responds to launches where the URL contains `vitalsource_book=true`. It uses the `VitalSourceService` service to generate the necessary configuration for the LMS frontend to load the VS viewer.
 - Add logic to the frontend in `BasicLtiLaunchApp` to check for this config and respond by rendering a `VitalSourceBookViewer` component instead of the normal Via iframe. The `VitalSourceBookViewer` component in turn renders an iframe containing a form with the LTI launch parameters and auto-submits the form. This results in the iframe then displaying the VitalSource book viewer
   
----

**Testing:**

1. Configure the `VITALSOURCE_LAUNCH_KEY` and `VITALSOURCE_LAUNCH_SECRET` env vars in your local lms dev environment. H staff can find the credentials in the "VitalSource LTI and API keys" entry in 1Password.
2. Create an assignment using the LMS app in Canvas and manually set the assignment's URL to http://localhost:8001/lti_launches?vitalsource_book=true&book_id=BOOKSHELF-TUTORIAL&cfi=%2F6%2F8%5B%3Bvnd.vst.idref%3Dvst-70a6f9d3-0932-45ba-a583-6060eab3e536%5D. Once https://github.com/hypothesis/lms/pull/2388 is merged, it will be possible to use the file picker UI to configure a VitalSource assignment
3. Launch the assignment. It should load the VitalSource book viewer app in an iframe instead of Via as for other assignments. I suggest clicking the option to open VS in a new window, since the book viewer is difficult to use in the iframe Canvas provides.
4. You may be assigned to enter VitalSource account details when the viewer first appears. Either create an account for yourself or ping me and I can give you credentials to use. You may see a "No, skip this step" option at the login screen, in which case just click that.
5. You may see a bunch of popups and other cruft in the viewer that need to be dismissed before you can get to the text. We are going to work with the VS team to remove this

After this, you should see the the viewer load a tutorial book at the following page:

<img width="1196" alt="vitalsource-book" src="https://user-images.githubusercontent.com/2458/106887618-52ecfa80-66dd-11eb-8386-ca2a2eaa554f.png">

You won't see the Hypothesis sidebar yet, but if you select some text you will see the Annotate/Highlight toolbar appear, although depending on where and in which direction you create the selection, it may be obscured by VitalSource's own annotation tools.
